examples/face_recognition_knn.py

```diff
@@ -0,0 +1,127 @@
+    :param verbose: verbosity of training
+    :return: returns knn classifier that was trained on the given data.
+    """
+    X = []
+    y = []
+    for class_dir in listdir(train_dir):
+        if not isdir(join(train_dir, class_dir)):
+            continue
+        for img_path in image_files_in_folder(join(train_dir, class_dir)):
+            image = face_recognition.load_image_file(img_path)
+            faces_bboxes = face_locations(image)
+            if len(faces_bboxes) != 1:
+                if verbose:
+                    print("image {} not fit for training: {}".format(img_path, "didn't find a face" if len(faces_bboxes) < 1 else "found more than one face"))
+                continue
+            X.append(face_recognition.face_encodings(image, known_face_locations=faces_bboxes)[0])
+            y.append(class_dir)
+
+
+    if n_neighbors is None:
+        n_neighbors = int(round(sqrt(len(X))))
+        if verbose:
+            print("Chose n_neighbors automatically as:", n_neighbors)
+
+    knn_clf = neighbors.KNeighborsClassifier(n_neighbors=n_neighbors, algorithm=knn_algo, weights='distance')
+    knn_clf.fit(X, y)
+
+    if model_save_path != "":
+        with open(model_save_path, 'wb') as f:
+            pickle.dump(knn_clf, f)
+    return knn_clf
+
+def predict(X_img_path, knn_clf = None, model_save_path ="", DIST_THRESH = .5):
+    """
+    recognizes faces in given image, based on a trained knn classifier
+
+    :param X_img_path: path to image to be recognized
+    :param knn_clf: (optional) a knn classifier object. if not specified, model_save_path must be specified.
+    :param model_save_path: (optional) path to a pickled knn classifier. if not specified, model_save_path must be knn_clf.
+    :param DIST_THRESH: (optional) distance threshold in knn classification. the larger it is, the more chance of misclassifying an unknown person to a known one.
+    :return: a list of names and face locations for the recognized faces in the image: [(name, bounding box), ...].
+        For faces of unrecognized persons, the name 'N/A' will be passed.
+    """
+
+    if not isfile(X_img_path) or splitext(X_img_path)[1][1:] not in ALLOWED_EXTENSIONS:
+        raise Exception("invalid image path: {}".format(X_img_path))
+
+    if knn_clf is None and model_save_path == "":
+        raise Exception("must supply knn classifier either thourgh knn_clf or model_save_path")
+
+    if knn_clf is None:
+        with open(model_save_path, 'rb') as f:
+            knn_clf = pickle.load(f)
+
+    X_img = face_recognition.load_image_file(X_img_path)
+    X_faces_loc = face_locations(X_img)
+    if len(X_faces_loc) == 0:
+        return []
+
+    faces_encodings = face_recognition.face_encodings(X_img, known_face_locations=X_faces_loc)
+
+
+    closest_distances = knn_clf.kneighbors(faces_encodings, n_neighbors=1)
+
+    is_recognized = [True if closest_distances[0][i][0] <= DIST_THRESH else False for i in range(len(X_faces_loc))]
```

Comment:
can be rewritten as:
`[closest_distances[0][i][0] <= DIST_THRESH  for i in range(len(X_faces_loc))]`